ipns/record.go

```diff
@@ -3,124 +3,466 @@ package ipns
+		return 1, nil
+	} else if !aHasV2Sig && bHasV2Sig {
+		return -1, nil
+	}
+
+	as, err := a.Sequence()
+	if err != nil {
+		return 0, err
+	}
+
+	bs, err := b.Sequence()
+	if err != nil {
+		return 0, err
+	}
+
+	if as > bs {
+		return 1, nil
+	} else if as < bs {
+		return -1, nil
 	}
 
-	return selectRecord(recs, vals)
+	at, err := a.Validity()
+	if err != nil {
+		return 0, err
+	}
+
+	bt, err := b.Validity()
+	if err != nil {
+		return 0, err
+	}
+
+	if at.After(bt) {
+		return 1, nil
+	} else if bt.After(at) {
+		return -1, nil
+	}
+
+	return 0, nil
 }
 
-func selectRecord(recs []*pb.IpnsEntry, vals [][]byte) (int, error) {
-	switch len(recs) {
-	case 0:
-		return -1, errors.New("no usable records in given set")
-	case 1:
-		return 0, nil
+// EmbedPublicKey embeds the given public key in the given [Record]. While not
+// strictly required, some notes (e.g., DHT servers), may reject IPNS Records
+// that do not embed their public keys as they may not be able to validate them
+// efficiently.
+func EmbedPublicKey(r *Record, pk ic.PubKey) error {
+	// Try extracting the public key from the ID. If we can, do not embed it.
+	pid, err := peer.IDFromPublicKey(pk)
+	if err != nil {
+		return err
 	}
+	if _, err := pid.ExtractPublicKey(); err != peer.ErrNoPublicKey {
+		// Either a *real* error or nil.
+		return err
+	}
+
+	// We failed to extract the public key from the peer ID, embed it.
+	pkBytes, err := ic.MarshalPublicKey(pk)
+	if err != nil {
+		return err
+	}
+
+	r.pb.PubKey = pkBytes
+	return nil
+}
 
-	var i int
-	for j := 1; j < len(recs); j++ {
-		cmp, err := Compare(recs[i], recs[j])
+// ExtractPublicKey extracts a [crypto.PubKey] matching the given [peer.ID] from
+// the IPNS Record, if possible.
+func ExtractPublicKey(r *Record, pid peer.ID) (ic.PubKey, error) {
+	if pk, err := r.PubKey(); err == nil {
+		expPid, err := peer.IDFromPublicKey(pk)
 		if err != nil {
-			return -1, err
+			return nil, multierr.Combine(ErrInvalidPublicKey, err)
 		}
-		if cmp == 0 {
-			cmp = bytes.Compare(vals[i], vals[j])
-		}
-		if cmp < 0 {
-			i = j
+
+		if pid != expPid {
+			return nil, ErrPublicKeyMismatch
 		}
+
+		return pk, nil
+	} else if !errors.Is(err, ErrPublicKeyNotFound) {
+		return nil, multierr.Combine(ErrInvalidPublicKey, err)
+	} else {
+		return pid.ExtractPublicKey()
+	}
+}
+
+// RoutingRecordKey returns the IPNS routing key for the given [peer.ID]. Note
+// that the intended use of this function is for routing only. The output of this
+// function is not human readable. For a human readable version, see [RecordKey].
+func RoutingRecordKey(pid peer.ID) string {
+	return "/ipns/" + string(pid)
+}
+
+// RecordKey returns the human-readable IPNS key for the given [peer.ID].
+func RecordKey(pid peer.ID) string {
+	cid := peer.ToCid(pid)
+	encoded, err := cid.StringOfBase(multibase.Base36)
+	if err != nil {
+		panic("cid.StringOfBase was called with wrong parameters")
 	}
 
-	return i, nil
+	return "/ipns/" + encoded
 }
```

Comment:
This will likely create confusion. `RecordKey` now returns the human readable record key. Before, it returned what the `RoutingRecordKey` now does return. Since the arguments are the same, it's hard to spot. I added it clearly in the Changelog. We are publishing a breaking change after all.

If someone else prefers to make this more safe, I would suggest just renaming to `Key(pid peer.ID)` and `RoutingKey(pid peer.ID)`. That would force people to look at the new options and see the difference.